### PR TITLE
Metrics protos

### PR DESCRIPTION
**Pull request by mdittmer**
**NOTE**: PR is WIP pending design review.

Drop previous protos (which were unused and may cause confusion). Protos in this PR are from the design outlined [here](https://docs.google.com/document/d/1MZRF3EdpMG0pLNI99jEj-pBITbqWKYMyJC7Ga8NtZ88/edit).

At the moment none of the protos attempt to be backwards-compatible with the ad hoc structures currently in Cloud Storage, but that could be changed if there is a strong desire to do so.

As the above-linked doc describes, these protos should be pressed into service soon to store data in BigQuery tables.

//cc @mattl @jeffcarp

### CODE REVIEW

**Comment by rwaldron** [comment]
@mdittmer thanks for writing that document and providing it as a reference for use by reviewers—excellent stuff :)

File: protos/test_result.proto [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  string os_version_str = 2;
+
+  Engine engine = 3;
+  Browser browser = 4;
+  string browser_version_str = 5;
+
+  google.protobuf.Timestamp test_run_start_time = 6;
+  google.protobuf.Timestamp test_run_end_time = 7;
+  string long_wpt_hash = 8;
+  google.protobuf.Timestamp wpt_commit_time = 9;
```
Comment (rwaldron) [inline protos/test_result.proto]: The `wpt` data should both be `wpt_*`, ie. `wpt_hash` & `wpt_commit_time` (although, I'm not sure that `wpt_commit_time` is even necessary)

File: protos/test_result.proto [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+// An result for a single test or sub-test from a TestRun.
+message TestResult {
+  option (gen_bq_schema.table_name) = "test_results";
+
+  OperatingSystem os = 1;
+  string os_version_str = 2;
+
+  Engine engine = 3;
+  Browser browser = 4;
+  string browser_version_str = 5;
```
Comment (rwaldron) [inline protos/test_result.proto]: Unless all fields are going to be suffixed with some data type indicator, I would suggest removing `_str`

File: protos/test_result.proto [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  google.protobuf.Timestamp test_run_start_time = 6;
+  google.protobuf.Timestamp test_run_end_time = 7;
+  string long_wpt_hash = 8;
+  google.protobuf.Timestamp wpt_commit_time = 9;
+
+  string test_name = 10;
+  TestStatus status = 11;
```
Comment (rwaldron) [inline protos/test_result.proto]: `status` => `test_status`

File: protos/test_result.proto [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  string result_message = 12;
```
Comment (rwaldron) [inline protos/test_result.proto]: `result_message` => `test_message`

File: protos/test_result.proto [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  string test_sub_name = 13;
+  SubTestStatus sub_status = 14;
```
Comment (rwaldron) [inline protos/test_result.proto]: Presently, the individual test result files that contain the "subtests" data produce JSON with the following shape: 

```json
{
  "test": "/shadow-dom/event-post-dispatch.html",
  "status": "ERROR",
  "message": "Object doesn't support property or method 'attachShadow'",
  "subtests": [
    {
      "status": "FAIL",
      "message": "assert_equals: expected 0 but got 3",
      "name": "Event properties post dispatch without ShadowRoots (composed: true)."
    }, {
      "status": "FAIL",
      "message": "assert_equals: expected 0 but got 3",
      "name": "Event properties post dispatch without ShadowRoots (composed: false)."
    }
  ]
}
```
(That's a real example https://storage.googleapis.com/wptd/5d55258739/edge-15-windows-10-sauce/shadow-dom/event-post-dispatch.html)

If I understand correctly, based on the design of `TestResult`, that data would look something like this: 


| os | os_version_str | engine | browser | browser_version_str | test_run_start_time | test_run_end_time | long_wpt_hash | wpt_commit_time | test_name | status | result_message | test_sub_name | sub_status | result_sub_message |
| -- | -------------- | ------ | ------- | ------------------- | ------------------- | ----------------- | ------------- | --------------- | --------- | ------ | -------------- | ------------- | ---------- | ------------------ |
| 3 | "10" | 2 | 2 | "15" |  test_run_start_time |  test_run_end_time | 5d552587396f6ee3148c4991a87130fbbdffd324 | wpt_commit_time | "/shadow-dom/event-post-dispatch.html" | 2 | "Object doesn't support property or method 'attachShadow'" | "Event properties post dispatch without ShadowRoots (composed: true)." | 2 | "assert_equals: expected 0 but got 3" |
| 3 | "10" | 2 | 2 | "15" |  test_run_start_time |  test_run_end_time | 5d552587396f6ee3148c4991a87130fbbdffd324 | wpt_commit_time | "/shadow-dom/event-post-dispatch.html" | 2 | "Object doesn't support property or method 'attachShadow'" | "Event properties post dispatch without ShadowRoots (composed: false)." | 2 | "assert_equals: expected 0 but got 3" |


Which now has duplicated the following information in both rows:

```js
  "test": "/shadow-dom/event-post-dispatch.html",
  "status": "ERROR",
  "message": "Object doesn't support property or method 'attachShadow'",
```

File: protos/test_run.proto [hunk 6]
```diff
@@ -16,20 +16,26 @@ syntax = "proto3";
...
 
+// An invocation of the WPT runner in some OS+browser context.
 message TestRun {
-  google.protobuf.Timestamp created_at = 1;
+  option (gen_bq_schema.table_name) = "test_runs";
 
-  // The first 10 characters of the SHA1 of the tested WPT revision
-  string revision = 2;
+  google.protobuf.Timestamp start_time = 1;
+  google.protobuf.Timestamp end_time = 2;
```
Comment (rwaldron) [inline protos/test_run.proto]: These fields are prefixed with `test_run_*` in `TestResult`

File: protos/test_result.proto [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+  google.protobuf.Timestamp test_run_start_time = 6;
+  google.protobuf.Timestamp test_run_end_time = 7;
+  string long_wpt_hash = 8;
+  google.protobuf.Timestamp wpt_commit_time = 9;
+
+  string test_name = 10;
+  TestStatus status = 11;
+  string result_message = 12;
+
```
Comment (rwaldron) [inline protos/test_result.proto]: Alternatively, they could all have the `test_` prefix removed entirely, which would match the current naming.

File: protos/test_run.proto [hunk 8]
```diff
@@ -16,20 +16,26 @@ syntax = "proto3";
...
+  string os_version_str = 4;
 
-  // Resource URL for results JSON blob
-  string results_url = 7;
+  Engine engine = 5;
+  Browser browser = 6;
+  string browser_version_str = 7;
+
+  string long_wpt_hash = 8;
+  google.protobuf.Timestamp wpt_commit_time = 9;
```
Comment (rwaldron) [inline protos/test_run.proto]: Same feedback as above: 

> The `wpt` data should both be `wpt_*`, ie. `wpt_hash` & `wpt_commit_time` (although, I'm not sure that `wpt_commit_time` is even necessary)

File: protos/test_run.proto [hunk 9]
```diff
@@ -16,20 +16,26 @@ syntax = "proto3";
...
-  string os_name = 5;
-  string os_version = 6;
+  OperatingSystem os = 3;
+  string os_version_str = 4;
 
-  // Resource URL for results JSON blob
-  string results_url = 7;
+  Engine engine = 5;
+  Browser browser = 6;
+  string browser_version_str = 7;
```
Comment (rwaldron) [inline protos/test_run.proto]: Same feedback as above:

> Unless all fields are going to be suffixed with some data type indicator, I would suggest removing `_str`

File: protos/test_status.proto [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+syntax = "proto3";
+
+package wptdashboard;
+
+enum TestStatus {
+  TEST_STATUS_UNKNOWN = 0;
+  TEST_OK = 1;
+  TEST_ERROR = 2;
+  TEST_TIMEOUT = 3;
```
Comment (rwaldron) [inline protos/test_status.proto]: Following the convention used in `Browser`, `Engine` and `OperatingSystem`: 

- `TEST_STATUS_UNKNOWN`
- `TEST_OK` => `OK`
- `TEST_ERROR` => `ERROR`
- `TEST_TIMEOUT` => `TIMEOUT`


Or alternatively: 

```go
enum TestStatus {
  UNKNOWN = 0;
  OK = 1;
  ERROR = 2;
  TIMEOUT = 3;
}
```

WDYT?

File: protos/sub_test_status.proto [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+syntax = "proto3";
+
+package wptdashboard;
+
+enum SubTestStatus {
+  SUB_TEST_STATUS_UNKNOWN = 0;
+  SUB_TEST_PASS = 1;
+  SUB_TEST_FAIL = 2;
+  SUB_TEST_TIMEOUT = 3;
+  SUB_TEST_NOT_RUN = 4;
```
Comment (rwaldron) [inline protos/sub_test_status.proto]: Following the convention used in `Browser`, `Engine` and `OperatingSystem`: 

- `SUB_TEST_STATUS_UNKNOWN`
- `SUB_TEST_PASS` => `PASS`
- `SUB_TEST_FAIL` => `FAIL`
- `SUB_TEST_TIMEOUT` => `TIMEOUT`
- `SUB_TEST_NOT_RUN` => `NOT_RUN`


Or alternatively: 

```go
enum SubTestStatus {
  UNKNOWN = 0;
  PASS = 1;
  FAIL = 2;
  TIMEOUT = 3;
  NOT_RUN = 4;
}
```

WDYT?

File: protos/browser.proto [hunk 12]
```diff
@@ -16,13 +16,10 @@ syntax = "proto3";
...
-  string browser_version = 4;
-  string os_name = 5;
-  string os_version = 6;
-  bool sauce = 7;
+enum Browser {
+  BROWSER_UNKNOWN = 0;
+  CHROME = 1;
+  EDGE = 2;
+  FIREFOX = 3;
+  SAFARI = 4;
```
Comment (rwaldron) [inline protos/browser.proto]:
```go
enum Browser {
  UNKNOWN = 0;
  CHROME = 1;
  EDGE = 2;
  FIREFOX = 3;
  SAFARI = 4;
}
```

WDYT?

File: protos/engine.proto [hunk 13]
```diff
@@ -16,6 +16,10 @@ syntax = "proto3";
...
 package wptdashboard;
 
-message Token {
-  string secret = 1;
+enum Engine {
+  ENGINE_UNKNOWN = 0;
+  CHROMIUM = 1;
+  EDGE_HTML = 2;
+  GECKO = 3;
+  WEB_KIT = 4;
```
[the remark this answers is not in the document]
Comment (rwaldron) [reply protos/engine.proto]:
```go
enum Engine {
  UNKNOWN = 0;
  BLINK = 1;
  EDGEHTML = 2;
  GECKO = 3;
  WEBKIT = 4;
}
```

WDYT?

File: protos/operating_system.proto [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+syntax = "proto3";
+
+package wptdashboard;
+
+enum OperatingSystem {
+  OPERATING_SYSTEM_UNKNOWN = 0;
+  DARWIN = 1;
+  LINUX = 2;
+  WINDOWS = 3;
```
Comment (rwaldron) [inline protos/operating_system.proto]:
```go
enum OperatingSystem {
  UNKNOWN = 0;
  DARWIN = 1;
  LINUX = 2;
  WINDOWS = 3;
}
```

WDYT?

File: protos/browser.proto (hunk 12 above)
Comment (mdittmer) [reply protos/browser.proto]: Without putting these in a different package, we can only use `UNKNOWN` once in an enum (the names get hoisted up into their package namespace a-la C/C++). I wish it weren't so; I would tend to agree with you.

File: protos/engine.proto [hunk 15]
```diff
@@ -16,6 +16,10 @@ syntax = "proto3";
 
 package wptdashboard;
 
-message Token {
-  string secret = 1;
+enum Engine {
+  ENGINE_UNKNOWN = 0;
+  CHROMIUM = 1;
```
[the remark this answers is not in the document]
Comment (mdittmer) [reply protos/engine.proto]: Of course! That was silly of me.

File: protos/engine.proto (hunk 13 above)
[the remark this answers is not in the document]
Comment (mdittmer) [reply protos/engine.proto]: ditto. I'm fine with the COMPOUNDWORD comment, though.

File: protos/operating_system.proto (hunk 14 above)
Comment (mdittmer) [reply protos/operating_system.proto]: ditto

File: protos/sub_test_status.proto (hunk 11 above)
Comment (mdittmer) [reply protos/sub_test_status.proto]: ditto (collisions with TestStatus on some non-`UNKNOWN` names too)

File: protos/test_result.proto [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+import "bq_table_name.proto";
+import "browser.proto";
+import "engine.proto";
+import "google/protobuf/timestamp.proto";
+import "operating_system.proto";
+import "sub_test_status.proto";
+import "test_status.proto";
+
+// An result for a single test or sub-test from a TestRun.
```
[the remark this answers is not in the document]
Comment (mdittmer) [reply protos/test_result.proto]: Done (next push).

File: protos/test_result.proto (hunk 2 above)
Comment (mdittmer) [reply protos/test_result.proto]: I don't feel too strongly about this, but I tried to add indicators where the representation might be ambiguous: "Did we store a string, or major/minor/patch numbers, or...?" WDYT of this rationale?

File: protos/test_result.proto (hunk 1 above)
Comment (mdittmer) [reply protos/test_result.proto]: SGTM. I'll add a comment that it's the full-length hash. The commit_time is intended to ease BigQuery queries concerned with commit time (without having to cross-reference to a checkout of WPT for every row).

File: protos/test_result.proto (hunk 7 above)
Comment (mdittmer) [reply protos/test_result.proto]: The thing that push the common prefix here is that `message` is a reserved word in proto3.

File: protos/test_result.proto (hunk 5 above)
Comment (mdittmer) [reply protos/test_result.proto]: Your assessment is exactly correct. The intent is that results can be filtered and/or grouped easily with a non-nested schema in BigQuery (and duplicate data is cheap).

File: protos/test_run.proto (hunk 6 above)
Comment (mdittmer) [reply protos/test_run.proto]: Yes. The intent there is to qualify that it is the "test run's start/end time" in the context of a TestResult; in the context of a TestRun, that seems verbose, IMHO.

File: protos/test_result.proto (hunk 5 above)
Comment (mdittmer) [reply protos/test_result.proto]: Put another way: We trade "wasted" storage for query simplicity (i.e., no need to join against a separate `sub_tests` table, or query over a `RECORD`/`STRUCT` field in BigQuery).

File: protos/browser.proto (hunk 12 above)
Comment (rwaldron) [reply protos/browser.proto]: > we can only use UNKNOWN once in an enum...
> ... (the names get hoisted up into their package namespace a-la C/C++

Well, at least it's consistent with _something_ ;) Thanks for the follow up

File: protos/sub_test_status.proto (hunk 11 above)
Comment (rwaldron) [reply protos/sub_test_status.proto]: Sorry about all this ditto-ing

File: protos/test_result.proto (hunk 2 above)
Comment (rwaldron) [reply protos/test_result.proto]: That's fine, this certainly wasn't a blocker.

File: protos/test_run.proto (hunk 6 above)
Comment (rwaldron) [reply protos/test_run.proto]: Ok, I see now and agree with your thinking—thanks for clarifying

File: protos/test_run.proto (hunk 9 above)
Comment (rwaldron) [reply protos/test_run.proto]: Strike this

File: protos/test_status.proto (hunk 10 above)
Comment (rwaldron) [reply protos/test_status.proto]: Ignore the second part

Comment (mdittmer) [reply protos/test_status.proto]: There's a name collision on `TIMEOUT` :cry:. That was the reason I aimed for a common-prefix strategy. WDYT?

File: protos/browser.proto (hunk 12 above)
Comment (lukebjerring) [reply protos/browser.proto]: From the Googleverse somewhere:
> The recommendation is to prefix the enum values with the enum’s name if there is a chance of name clashes

Ugly, but there it is

File: protos/browser_specific_pass_rate_summary.proto [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+// share "base_name" prefix in their TestResult.test_name.
+message BrowserSpecificPassRateSummary {
+  option (gen_bq_schema.table_name) = "browser_specific_pass_rate_summaries";
+
+  OperatingSystem os = 1;
+  string os_version_str = 2;
+
+  Engine engine = 3;
+  Browser browser = 4;
+  string browser_version_str = 5;
```
Comment (lukebjerring) [inline protos/browser_specific_pass_rate_summary.proto]: I'd probably aggregate these into 'platform', which can have a _large_ number of entries, and shard into these 3 enums as metadata (if/when needed)

Comment (mdittmer) [reply protos/browser_specific_pass_rate_summary.proto]: I moved `Engine` to be `Browser` metadata, but I think we may want to query against `Browser` and `OperatingSystem` separately sometimes.

**Comment by rwaldron** [comment]
A few things here: 

1. I messed up the order in which the metrics stuff was landed and the second PR needs to be backed out before we proceed. I have requested reality checks on branch that does that. So far @lukebjerring has responded with confirmation. I'm waiting for at least one more confirmation. 
2. Landing this branch will have an impact on the work in https://github.com/w3c/wptdashboard/pull/315, is there a plan? If the plan is simply: "get it resolved and move the files to where they belong", that's perfectly fine too.

**Comment by rwaldron** [comment]
```
git remote add upstream git@github.com:w3c/wptdashboard.git;
git checkout metrics-protos;
git pull --rebase upstream master;
git push upstream HEAD -f;
```

Here's what I hit: 

> CONFLICT (modify/delete): protos/browser.proto deleted in Drop previous browser.proto and modified in HEAD. Version HEAD of protos/browser.proto left in tree.

Solution: 

```
git rm protos/browser.proto
git rebase --continue
```

> CONFLICT (modify/delete): protos/token.proto deleted in Delete old Token proto and modified in HEAD. Version HEAD of protos/token.proto left in tree.

Solution: 

```
git rm protos/token.proto
git rebase --continue
```

> CONFLICT (content): Merge conflict in Makefile

Solution: I resolved with "theirs" (as in: the changes in _this_branch) and again `git rebase --continue` which will apply the last two commits and give a clean, rebased version of this branch. 

OR... if you're too annoyed with me for all of this churn, I've done it for you: https://github.com/rwaldron/wptdashboard/tree/metrics-protos ;)

**Comment by rwaldron** [comment]
@mdittmer is this PR still active?

**Pull request closed without merging**